state/src/impls/top_level.rs

```diff
@@ -356,6 +344,46 @@ impl TopLevelState {
         result
     }
 
+    fn get_fee_payer(&self, signer_public: &Public, key_allowance: KeyAllowance) -> StateResult<Address> {
```

Comment:
Extracting `get_fee_payer` looks good, but introducing KeyAllowance reduces the readability.